File: src/player/gain.rs
//! BS.1770 playback normalization policy.
//!
//! Port of `web/player-core/src/gain.ts` (BSD-3-Clause). `.mpack` values are
//! never modified: this derives a playback gain only.
//!
//! Phase 9's [`crate::audio::gain_db`] is the low-level measurement
//! primitive (`target − measured`); this module is the player's
//! normalization *policy* layered on top of it (mode selection, the true-peak
//! ceiling, and the linear combination with user volume).

use super::types::{AlbumLoudness, TrackLoudness};

/// Initial client playback target (documented playback policy).
pub const PLAYBACK_TARGET_LUFS: f64 = -16.0;

/// Ceiling applied to the normalized true peak (headroom for downstream).
pub const TRUE_PEAK_CAP_DB: f64 = -1.0;

/// Normalization gain in dB for a track under the given mode.
///
/// Missing loudness (or mode `off`) yields `0`. The gain is
/// `target − measured`, reduced when necessary so the output true peak stays
/// at or below [`TRUE_PEAK_CAP_DB`].
pub fn normalization_gain(
    mode: super::types::NormalizationMode,
    track: Option<&TrackLoudness>,
    album: Option<&AlbumLoudness>,
) -> f64 {
    use super::types::NormalizationMode;
    if mode == NormalizationMode::Off {
        return 0.0;
    }
    let measured = match mode {
        NormalizationMode::Album => album.map(|a| a.album_lufs),
        NormalizationMode::Track => track.map(|t| t.lufs),
        NormalizationMode::Off => None,
    };
    let peak = match mode {
        NormalizationMode::Album => album.map(|a| a.album_true_peak_db),
        NormalizationMode::Track => track.map(|t| t.true_peak_db),
        NormalizationMode::Off => None,
    };
    let Some(measured) = measured else {
        return 0.0;
    };
    let mut gain = PLAYBACK_TARGET_LUFS - measured;
    if let Some(peak) = peak {
        let max_gain = TRUE_PEAK_CAP_DB - peak;
        if gain > max_gain {
            gain = max_gain;
        }
    }
    gain
}

/// Converts dB to a linear gain factor (`10^(db/20)`).
pub fn db_to_linear(db: f64) -> f64 {
    10f64.powf(db / 20.0)
}

/// Combined linear gain = user volume (0..1) × normalization gain.
pub fn combined_gain(user_volume: f64, norm_db: f64) -> f64 {
    user_volume * db_to_linear(norm_db)
}

File: web/player-core/src/gain.ts
// Copyright (c) 2026, The MusicPack Development Team
// SPDX-License-Identifier: BSD-3-Clause

// BS.1770 playback normalization (client policy — .mpack values are never
// modified). gain = playback target - measured loudness, constrained so the
// output true peak never exceeds TRUE_PEAK_CAP_DB.

export type NormalizationMode = 'off' | 'album' | 'track';

/** Initial client playback target (documented playback policy). */
export const PLAYBACK_TARGET_LUFS = -16;
/** Ceiling applied to the normalized true peak (headroom for downstream). */
export const TRUE_PEAK_CAP_DB = -1;

export interface TrackLoudnessLike {
  lufs: number;
  truePeakDb: number;
}

export interface AlbumLoudnessLike {
  albumLufs: number;
  albumTruePeakDb: number;
}

/** Normalization gain in dB for a track under the given mode. */
export function normalizationGainDb(
  mode: NormalizationMode,
  track?: TrackLoudnessLike,
  album?: AlbumLoudnessLike,
): number {
  if (mode === 'off') return 0;
  const measured = mode === 'album' ? album?.albumLufs : track?.lufs;
  const peak = mode === 'album' ? album?.albumTruePeakDb : track?.truePeakDb;
  if (measured === undefined) return 0;
  let gain = PLAYBACK_TARGET_LUFS - measured;
  if (peak !== undefined) {
    const maxGain = TRUE_PEAK_CAP_DB - peak;
    if (gain > maxGain) gain = maxGain;
  }
  return gain;
}

export function dbToLinear(db: number): number {
  return Math.pow(10, db / 20);
}

/** Combined linear gain = user volume (0..1) x normalization gain. */
export function combinedGain(userVolume: number, normDb: number): number {
  return userVolume * dbToLinear(normDb);
}
